decompiler_agent/src/main/java/org/jrd/agent/InstrumentationProvider.java

```diff
@@ -23,6 +23,20 @@
     InstrumentationProvider(Instrumentation inst, Transformer transformer) {
         this.transformer = transformer;
         this.instrumentation = inst;
+    }    
+        
+    public void setClassBody(String cname, byte[] nwBody) throws UnmodifiableClassException {
+        Class clazz = findClass(cname);
+        transformer.allowToSaveBytecode(); //?
+        try {
+            String nameWithSlashes = clazz.getName().replace(".", "/");
+            transformer.allowToSaveBytecode(); //? taken from getClassBody. Really twice?
```

Comment:
Why is that boolean even needed? Why not always allow saving bytecode? I think you can just delete it and it shouldn't break anything.

Reply:
I will ask Alice for this. I believe she had osme performance issues here when reading that al lthe time,

Reply:
Hi,
I really don't know why it is there twice, probably a leftover from long coding night, switching it on once should be enough. Please test it, if you change it, just in case, this is where the JVM dark magic begins.

The boolean switch however has its purpose there, the idea was given to me by Andrew Dinn (and later figured out by Jiri actually), when I was resolving the issue with class reloading. 

When you look in the Main, premain method, the only thing this agent actually does is register a transformer, create InstrumentationProvider instance and runs the thread that listens on the port.

While the InstrumentationProvider is simply performing the methods on given instrumentation and with transformer t, when we ask him to do so, once we register the Transformer, it is in the JVM for the infinity, or unless we unregister it, but that is harder than the existing solution, and in general, we want to avoid doing that multiple times.

What we do instead is, we change the transformer behavior on the run, with the switch. Why? Because otherwise, every time the transform is called, the bytecode would be saved. 

That is probably fine in small apps, but remember, the transform methods of all registered transformers is called every time any Class is loaded into JVM, or retransformed. That means, when you load some bigger library into the JVM with our agent, that is hell a lot of saving for nothing and it could cause serious performance issues. The Instrumentation library has in general a different purpose, it should not have on and off switch. This is abusing the functionality to get the bytecode, since there is no other (known) way.

Another thing is, that we currently save the results into a map. That has to be done this way, because think of scenario, where you want bytecode of class A and save it into simple byte[]. But while we use the switch and save the bytecode, another class is loaded into the JVM, while the switch is still on. The results would get replaced, because the transform would save another class into the array. So we have map with class names, so even if we get 5 results in the moment, we get always the right class. In case where the switch is not there, we have map with bytecode of all classes ever loaded in the classloader. In bigger application, this would be very memory consuming.

In general, the switch is not the only possible solution, it is just something that came up, if you have more elegant solution, you can change it, but it definitely can't save a byte array of every class that was ever loaded.